src/dstruct.d

```diff
@@ -238,22 +238,27 @@ extern (C++) class StructDeclaration : AggregateDeclaration
     // For those, today TypeInfo_Struct is generated in COMDAT.
     bool requestTypeInfo;
 
-    final extern (D) this(Loc loc, Identifier id)
+    final extern (D) this(Loc loc, Identifier id, bool inObject)
     {
         super(loc, id);
         zeroInit = 0; // assume false until we do semantic processing
         ispod = ISPODfwd;
+
         // For forward references
         type = new TypeStruct(this);
-        if (id == Id.ModuleInfo && !Module.moduleinfo)
-            Module.moduleinfo = this;
+
+        if (inObject)
+        {
+            if (id == Id.ModuleInfo && !Module.moduleinfo)
```

Comment:
All the stuff in [dclass](https://github.com/dlang/dmd/blob/b7d84f8c7358d4bdf5518392bcf99831c09faf3b/src/dclass.d#L262) disallows using "reserved" classnames from object.d, should we deprecate other `struct ModuleInfo`s as well?